utils/psicc/psicc.1
.\"Shiju P. Nair September 30, 2004
.\"Thomas Weber <tweber@debian.org> April 23, 2014
.TH PSICC 1 "September 30, 2004"
.SH NAME
psicc - little cms PostScript converter.
.SH SYNOPSIS
.B psicc
.RI [ options ]
.SH DESCRIPTION
lcms is a standalone CMM engine, which deals with the color management.
It implements a fast transformation between ICC profiles.
.B psicc
is a little cms PostScript converter.
.SH OPTIONS
.TP
.B \-b
Black point compensation (CRD only).
.TP
.BI \-c\  precision
Precision (0=LowRes, 1=Normal, 2=Hi-res) (CRD only) [defaults to 1].
.TP
.BI \-i\  profile
Input profile: Generates Color Space Array (CSA).
.TP
.BI \-n\  gridpoints
Alternate way to set precision, number of CLUT points (CRD only).
.TP
.BI \-o\  profile
Output profile: Generates Color Rendering Dictionary(CRD).
.TP
.BI \-t\  intent
Intent (0=Perceptual, 1=Colorimetric, 2=Saturation, 3=Absolute) [defaults to 0].
.TP
.B \-u
Do NOT generate resource name on CRD.
.SH NOTES
For suggestions, comments, bug reports etc. send mail to
info@littlecms.com.
.SH SEE ALSO
.BR jpgicc (1),
.BR linkicc (1),
.BR tificc (1),
.BR transicc (1)
.SH AUTHOR
This manual page was written by Shiju p. Nair <shiju.p@gmail.com>,
for the Debian project.


ChangeLog
-----------------------
2.15 Mainenance release
-----------------------
New MESON build system, many thanks to amispark and Lovell Fuller for bringing this.       
Fixed a bug that caused memory corruption on colord
cmsReadRawTag can read portions  of tags again. Removing this caused colord to segfault when dumping profiles
Added more checks based of fuzzer discoveries.
MSYS2 can now compile lcms2
Checked on Apple Silicon M1 and M2 
Fixed a bug of fastfloat plug-in that affected Krita CMYK color selector

-----------------------
2.14 Featured release
-----------------------
lcms2 now implements ICC specification 4.4
New multi-threaded plug-in
several fixes to keep fuzzers happy
Remove check on DLL when CMS_NO_REGISTER_KEYWORD is used 
Added more validation against broken profiles
Add more help to several tools

-----------------------
2.13.1 Hot fix
-----------------------
Fix for pure white going gray in grayscale transforms.

-----------------------
2.13 Featured release
-----------------------
Added support for premultiplied alpha
tifficc can now handle alpha channels, both unassociated and premultiplied
Better documentation
CGATS parser can now deal with very long strings
Added Projects for Visual Studio 2020
Travis CI discontinued, GitHub actions used instead
Added a very preliminar meson build script (thanks to xclaesse)
Added ARM64 target to visual studio 2019 (thanks to gaborkertesz-linaro)
Added thread safe code to get time
Added automatic linear space detection 
Added cmsGetStageContextID function
Added cmsDetectRGBProfileGamma function
configure now accepts --without-fastfloat to turn plugin off
autogen.sh has now a --distclean toggle to get rid of all autotools generated files
Checked to work on STM32 Cortex-A, Cortex-M families
Bug & typos fixing (thanks to many reporters and contributors)

-----------------------
2.12 Maintenance release
-----------------------
Added new build-in sigmoidal tone curve
Added XCode 12 project
Added support for multichannel input up to 15 channels
Fix LUT8 write matrix
Fix version mess on 10/11
Fix tools & samples xgetopt
Fix warnings on different function pointers
Fix matlab MEX compilation
plugin: cleanup and better SSE detection
plugin: add lab to any on float
plugin: it can now be compiled as C++
recover PDF documentation, but try to keep it under a reasonable size.
Prevent a rare but possible out-of-bounds read in postscript generator
Remove unused variables

-----------------------
2.11 Maintenance release
-----------------------
Fixed __cpuid() on fast float plugin to allow gnu gcc
Fixed copy alpha bounds check
Fixed data race condition on contexts pool
Fixed LUT16 write matrix on multichannel V2 profiles

-----------------------
2.10 Featured release
-----------------------
Added a compilation toggle to remove "register" keyword in API. 
Previously commercial, fast_float plug-in is now released as open source under GPL3 license.
MD5 functions are now accessible through plug-in API.
Added support for Visual Studio 2019
Bug fixing.


-----------------------
2.9 Maintenance release
-----------------------
Several fixes related with security, and therefore not detailed here.
C++ compiles now without warnings
Added OSX and clang in travis
Added a travis-ci test matrix for many compilers and OS. Thanks to Thomas Weber (debian) for this great improvement.
testbed can now work with dynamic versions of library 
Fixed wrong planar formatters regarding linestride interpretation

-----------------------
2.8 Featured release
-----------------------

Changed ChangeLog direction
Fixed many typos in comments, thanks to Stefan Weil for doing that.
Fixed localization bug, added a new test case crayons.icc thanks to Richard Hughes for providing the profile. 
Fixed a bug in optimizer that made some formats (i.e, bits planar) unavailable
Fixed misalignment problems on Alpha. The compiler does not align strings, and accessing begin of string as a uint16 makes code to fail.
Added some extra checks to the tools and examples.
Fixed a bug that prevented to read luminance tag
BIG amount of functionality contributed/Sponsored  by Alien Skin Software: TransformStride, copyAlpha, performance plug-ins. Fixes some warnings as well.
Added an extra _ to _stdcall to make it more portable
Fixed a bug in transicc for named color profiles
Fixed several compiler warnings
Added support for Visual Studio 2015
Fixed for XCODE project

-----------------------
2.7 Maintenance release
-----------------------

Added a version retrieval function 
Added an option in transicc for working in bounded mode
Fixed wrong handling of extra channels in some formatters.
Added a project for VS2013
Added license for iccjpeg.c
New project for mac
Added a global optimization that merges consecutive matrices in pipelines. Fixes loss of precision in some transforms
Added a flag  to clip negative values in unbounded transforms (only gray, rgb, cmyk)
Move unused var suppresor before the `return` statements.
Remove dead code.
Add missing comma in CGATS parser    
utils/jpgicc/iccjpeg.c: Fix check if unsigned variable 'total_length'… …
Some maintenance fixes
Remove unused vcproj files
Added a function to retrieve the iohandler of a given profile object
Added a safety check on named color lists
Fixed a macro clash on SNONE. 
Fixed a possible segmentation fault in a non-happy path

-----------------------
2.6 Featured release
-----------------------

Added pthread dependency. From now lcms supports multithreading
Fix for delete tag memory corruption
Added directories for tiff, jpeg in configure script
New locking plug-in, from Artifex
Big revamp on Contexts, from Artifex
Fixed memory leaks on error handling
Changed endianness detection for PowerPC
Added a way to retrieve matrix shaper always, no matter LUT is present
Fixed a bug in PCS/Colorspace order when reading V2 Lab devicelinks
Fixed some indexing out of bounds in floating point interpolation

utils/psicc/psicc.c
//---------------------------------------------------------------------------------
//
//  Little Color Management System
//  Copyright (c) 1998-2022 Marti Maria Saguer
//
// Permission is hereby granted, free of charge, to any person obtaining 
// a copy of this software and associated documentation files (the "Software"), 
// to deal in the Software without restriction, including without limitation 
// the rights to use, copy, modify, merge, publish, distribute, sublicense, 
// and/or sell copies of the Software, and to permit persons to whom the Software 
// is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in 
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO 
// THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND 
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE 
// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION 
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION 
// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//
//---------------------------------------------------------------------------------

#include "utils.h"

// ------------------------------------------------------------------------

static char *cInProf = NULL;
static char *cOutProf = NULL;
static int Intent = INTENT_PERCEPTUAL;
static FILE* OutFile;
static int BlackPointCompensation = FALSE;
static int Undecorated = FALSE;
static int PrecalcMode = 1;
static int NumOfGridPoints = 0;



static
void Help(void)
{
    fprintf(stderr, "usage: psicc [flags] [<Output file>]\n\n");

    fprintf(stderr, "flags:\n\n");

    fprintf(stderr, "-i<profile> - Input profile: Generates Color Space Array (CSA)\n");
    fprintf(stderr, "-o<profile> - Output profile: Generates Color Rendering Dictionary(CRD)\n");

    fprintf(stderr, "-t<0,1,2,3> - Intent (0=Perceptual, 1=Colorimetric, 2=Saturation, 3=Absolute)\n");

    fprintf(stderr, "-b - Black point compensation (CRD only)\n");
    fprintf(stderr, "-u - Do NOT generate resource name on CRD\n");
    fprintf(stderr, "-c<0,1,2> - Precision (0=LowRes, 1=Normal (default), 2=Hi-res) (CRD only)\n");
    fprintf(stderr, "-n<gridpoints> - Alternate way to set precision, number of CLUT points (CRD only)\n");

    fprintf(stderr, "\n");
    fprintf(stderr, "If no output file is specified, output goes to stdout.\n\n");
    fprintf(stderr, "This program is intended to be a demo of the little cms\n"
        "engine. Both lcms and this program are freeware. You can\n"
        "obtain both in source code at https://www.littlecms.com\n"
        "For suggestions, comments, bug reports etc. send mail to\n"
        "info@littlecms.com\n\n");

    exit(0);
}

// The toggles stuff

static
void HandleSwitches(int argc, char *argv[])
{
       int s;
      
       while ((s = xgetopt(argc,argv,"uUbBI:i:O:o:T:t:c:C:n:N:-:")) != EOF) {

       switch (s) 
       {

       case '-':
           if (strcmp(xoptarg, "help") == 0)
           {
               Help();
           }
           else
           {
               FatalError("Unknown option - run without args to see valid ones.\n");
           }
           break;
	 
       case 'i':
       case 'I':
            cInProf = xoptarg;
            break;

       case 'o':
       case 'O':
           cOutProf = xoptarg;
            break;

       case 'b':
       case 'B': BlackPointCompensation =TRUE;
            break;


       case 't':
       case 'T':
            Intent = atoi(xoptarg);
            if (Intent > 3) Intent = 3;
            if (Intent < 0) Intent = 0;
            break;
     
       case 'U':
       case 'u':
            Undecorated = TRUE;
            break;

       case 'c':
       case 'C':
            PrecalcMode = atoi(xoptarg);
            if (PrecalcMode < 0 || PrecalcMode > 2)
                    FatalError("ERROR: Unknown precalc mode '%d'", PrecalcMode);
            break;


       case 'n':
       case 'N':
                if (PrecalcMode != 1)
                    FatalError("Precalc mode already specified");
                NumOfGridPoints = atoi(xoptarg);
                break;


  default:

       FatalError("Unknown option - run without args to see valid ones.\n");
    }       
    }
}




static
void GenerateCSA(void)
{
	cmsHPROFILE hProfile = OpenStockProfile(0, cInProf);
	size_t n;
	char* Buffer;

	if (hProfile == NULL) return;

	n = cmsGetPostScriptCSA(0, hProfile, Intent, 0, NULL, 0);
	if (n == 0) return;

    Buffer = (char*) malloc(n + 1);
    if (Buffer != NULL) {

        cmsGetPostScriptCSA(0, hProfile, Intent, 0, Buffer, (cmsUInt32Number) n);
        Buffer[n] = 0;

        fprintf(OutFile, "%s", Buffer);	

        free(Buffer);
    }

	cmsCloseProfile(hProfile);
}


static
void GenerateCRD(void)
{
	cmsHPROFILE hProfile = OpenStockProfile(0, cOutProf);
	size_t n;
	char* Buffer;
    cmsUInt32Number dwFlags = 0;
    
	if (hProfile == NULL) return;

    if (BlackPointCompensation) dwFlags |= cmsFLAGS_BLACKPOINTCOMPENSATION;
    if (Undecorated)            dwFlags |= cmsFLAGS_NODEFAULTRESOURCEDEF;

    switch (PrecalcMode) {
           	
	    case 0: dwFlags |= cmsFLAGS_LOWRESPRECALC; break;
		case 2: dwFlags |= cmsFLAGS_HIGHRESPRECALC; break;
		case 1: 
            if (NumOfGridPoints > 0)
                dwFlags |= cmsFLAGS_GRIDPOINTS(NumOfGridPoints);
            break;

		default: FatalError("ERROR: Unknown precalculation mode '%d'", PrecalcMode);
	 }

	n = cmsGetPostScriptCRD(0, hProfile, Intent, dwFlags, NULL, 0);
	if (n == 0) return;

	Buffer = (char*) malloc(n + 1);
	if (Buffer == NULL) return;
        cmsGetPostScriptCRD(0, hProfile, Intent, dwFlags, Buffer, (cmsUInt32Number) n);
	Buffer[n] = 0;

	fprintf(OutFile, "%s", Buffer);			
	free(Buffer);
	cmsCloseProfile(hProfile);
}


int main(int argc, char *argv[])
{
	int nargs;

    fprintf(stderr, "Little CMS ICC PostScript generator - v2.1 [LittleCMS %2.2f]\n", cmsGetEncodedCMMversion() / 1000.0);
    fprintf(stderr, "Copyright (c) 1998-2022 Marti Maria Saguer. See COPYING file for details.\n");
    fflush(stderr);

	// Initialize
	InitUtils("psicc");

	 HandleSwitches(argc, argv);

     nargs = (argc - xoptind);
	 if (nargs != 0 && nargs != 1)
				Help();            
	
    if (cInProf == NULL && cOutProf == NULL)
        Help();

	 if (nargs == 0) 
			OutFile = stdout;
	 else
			OutFile = fopen(argv[xoptind], "wt");
	   		
	  if (cInProf != NULL)
			GenerateCSA();
		  
	  if (cOutProf != NULL)
			GenerateCRD();
		
	  if (nargs == 1) {
		  fclose(OutFile);
	  }

      return 0;     
}


